dotnet/src/Agents/Framework/Internal/BroadcastQueue.cs

```diff
@@ -0,0 +1,140 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using ChannelQueue = System.Collections.Concurrent.ConcurrentQueue<System.Collections.Generic.IList<Microsoft.SemanticKernel.ChatMessageContent>>;
+
+namespace Microsoft.SemanticKernel.Agents.Internal;
+
+/// <summary>
+/// Utility class used by <see cref="AgentNexus"/> to manage the broadcast of
+/// conversation messages via the <see cref="AgentChannel"/>.
+/// (<see cref="AgentChannel.ReceiveAsync(IEnumerable{ChatMessageContent}, System.Threading.CancellationToken)"/>.)
+/// </summary>
+/// <remarks>
+/// Maintains a set of channel specific queues.
+/// </remarks>
+internal sealed class BroadcastQueue
+{
+    private int _isActive;
+    private readonly Dictionary<string, ChannelQueue> _queues = new();
+    private readonly Dictionary<string, Task> _tasks = new();
+    private readonly Dictionary<string, Exception> _failures = new();
+    private readonly object _queueLock = new(); // Synchronize access to _isActive, _queue and _tasks.
+
+    /// <summary>
+    /// Defines the yield duration when blocking for a channel-queue.
+    /// to drain.
+    /// </summary>
+    public TimeSpan BlockDuration { get; set; } = TimeSpan.FromSeconds(1);
+
+    /// <summary>
+    /// Enqueue a set of messages for a given channel.
+    /// </summary>
+    /// <param name="channels">The target channels for which to broadcast.</param>
+    /// <param name="messages">The messages being broadcast.</param>
+    public void Enqueue(IEnumerable<ChannelReference> channels, IList<ChatMessageContent> messages)
+    {
+        lock (this._queueLock)
```

Comment:
Has any alternative no-lock design (e.g., using concurrent dictionaries) been considered? Asking because of the potential problems with locks that are usually very difficult to debug, such as deadlocks and race conditions.

Reply:
Yes...in fact in the latest PR I've backed away from using ConcurrentQueue.  The reason is that the locking semantics are at a higher granularity that a single container or single invocation.  (Race management.)  

I'd love to spend some time with your paired up going through the latest version.

Reply:
Scheduled an hour on April 8th for us to dig into this.